### PR TITLE
Feature/goals

### PR DESCRIPTION
**Pull request by peerfikhman**
Hey Guys,
Appreciate if you check out my code and do a CR before accepting the PR changes.

Changes I've done:

1. Added goals field which is an array of moongose objectId to experiment schema with proper validation
2. Created a schema and exported model for goal mongoDB document with proper validations
3. Creating the goal repository which inherits basic repository functionality from MongoStorage and adds a method of getting success count
4. Made changes to the goal controller to export the goals functionality to the goal router
5. Made changes to the goal router to export functionality to API endpoints
6. Add to the experiment repository the functionality of adding and removing goals and exporting this functionality to the experiment controller and router
7.  Added populate functionality for goals on experiment

*** Please note that I have faced issues running the server due to the "newrelic" feature lately added and I would like your guidance about the manner. Meaning that the functionality added haven't been yet tested with postman.

### CODE REVIEW

File: controller/goal.controller.js [hunk 1]
```diff
@@ -28,5 +29,54 @@ module.exports = {
             res.status(200).send(await ExperimentRepository.getVariantSuccessCount(id));
         else
             throw new PropertyNotFound("getVariantSuccessCountByExperimentID");
+    },
+
+    getAllGoals: async (req, res) => {
+        res.status(200).send(await GoalRepository.find());
+    },
```
Comment (AtomicManic) [inline controller/goal.controller.js:36]: Do we need to get every goal in db? preferable to get goals by experiment id

File: models/Goal.model.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+const userSchema = new Schema(
+    {
+        name:{
+            type: String,
+            required: true
+        },
+        key : {
+            type : String,
+            required: true,
+        },
```
Comment (AtomicManic) [inline models/Goal.model.js:13]: what is the meaning of key for goal?

File: repositories/experiment.repository.js [hunk 3]
```diff
@@ -43,6 +66,14 @@ module.exports = new (class ExperimentsRepository extends MongoStorage {
...
+  }
+
+  async addGoal(experimentId, goalId){
+    return await this.update(experimentId, {$push: {goals: goalId}}).populate({path: 'goals'});
+  }
+  async removeGoal(experimentId, goalId){
+    return await this.update(experimentId, {$pull: {goals: goalId}}).populate({path: 'goals'});
   }
+
 })();
```
Comment (AtomicManic) [inline repositories/experiment.repository.js:79]: Great update on the experiment repository!

File: repositories/goal.repository.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+const MongoStorage = require("../db/mongo.storage");
+
+module.exports = new (class GoalRepository extends MongoStorage {
+    constructor() {
+        super("goal");
+    }
+
+    getGoalSuccessCountById(goalId) {
+        return this.retrieve(goalId)?.success_count ?? null;
```
Comment (AtomicManic) [inline repositories/goal.repository.js:9]: Could you explain the use of ?? operator? I know that it checks of the left operand is nullis, and if so, returns the right operand. If not, returns the left operand. So, if success count is null by itself, you could just return it.

File: repositories/goal.repository.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    }
+    incGoalSuccessCount(goalId){
+        return this.update(goalId, {$inc : {success_count: 1}})
+    }
+})();
```
Comment (AtomicManic) [inline repositories/goal.repository.js:14]: Great!

**Review by AtomicManic — COMMENTED** [review]
Hi. I left you some comments regarding the usability of the code, all the rest looks good. LMK about those comments and i'll approve the PR

File: controller/goal.controller.js (hunk 1 above)
Comment (peerfikhman) [reply controller/goal.controller.js:36]: You are right, It's not a must and can be deleted on the approval of the PR.
Retrieving a single goal by Id is supported as well

File: models/Goal.model.js (hunk 2 above)
Comment (peerfikhman) [reply models/Goal.model.js:13]: This is a field entered by the client to then access the specific goal once there are multiple goals.
for example, a client can have several goals with a goal where the key equals "login" stating that the end user had reached the login page, and another goal that has the key "purchase" for confirming a purchase

File: repositories/experiment.repository.js (hunk 3 above)
Comment (peerfikhman) [reply repositories/experiment.repository.js:79]: Thanks :)

File: repositories/goal.repository.js (hunk 4 above)
Comment (peerfikhman) [reply repositories/goal.repository.js:9]: It is also used for undefined in this case, making sure that return value is either a real value or null but not undefined, as we used as a convention in other parts of the code

File: repositories/goal.repository.js (hunk 5 above)
Comment (peerfikhman) [reply repositories/goal.repository.js:14]: Thank you 💯

File: models/Goal.model.js (hunk 2 above)
Comment (AtomicManic) [reply models/Goal.model.js:13]: So you are creating a goal groups of a sort? Or is it meant to be uniqe?

File: controller/goal.controller.js (hunk 1 above)
Comment (AtomicManic) [reply controller/goal.controller.js:36]: Great, so please remove this =]

File: repositories/goal.repository.js (hunk 4 above)
Comment (AtomicManic) [reply repositories/goal.repository.js:9]: thank for clarifying, great job

File: models/Goal.model.js (hunk 2 above)
Comment (peerfikhman) [reply models/Goal.model.js:13]: You are right it is supposed to be unique value, made the changes.

File: controller/goal.controller.js (hunk 1 above)
Comment (peerfikhman) [reply controller/goal.controller.js]: Removed

**Review by AtomicManic — APPROVED** [review]
Awesome work, approved

**Pull request merged**